### PR TITLE
add mode() function & tests

### PR DESCRIPTION
**Pull request by agaurav**
###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass
- [ ] CHANGELOG.md updated
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
 
Build result - https://circleci.com/gh/agaurav/influxdb/2
Resolves https://github.com/influxdata/influxdb/issues/1823

Also added bencmark test of the median function for comparison.
Benchmark Results on MacBook Air (2013) running Yosemite
```
PASS
BenchmarkQuery_String-4         	  200000	      7378 ns/op
BenchmarkCountIterator_1K-4     	   30000	     53432 ns/op	     832 B/op	      11 allocs/op
BenchmarkCountIterator_100K-4   	     300	   5011769 ns/op	     849 B/op	      11 allocs/op
BenchmarkCountIterator_1M-4     	      30	  51133687 ns/op	    1005 B/op	      11 allocs/op
BenchmarkDistinctIterator_1K-4  	   20000	     98592 ns/op	    7577 B/op	      18 allocs/op
BenchmarkDistinctIterator_100K-4	     200	   8983528 ns/op	    7582 B/op	      18 allocs/op
BenchmarkDistinctIterator_1M-4  	      20	  86981320 ns/op	    7560 B/op	      18 allocs/op
BenchmarkMedianIterator_1K-4    	    5000	    220252 ns/op	  197299 B/op	      22 allocs/op
BenchmarkMedianIterator_100K-4  	       5	 212307202 ns/op	56370470 B/op	     105 allocs/op
BenchmarkMedianIterator_1M-4    	       1	2336160091 ns/op	529295104 B/op	     158 allocs/op
BenchmarkModeIterator_1K-4      	   10000	    153614 ns/op	  197299 B/op	      22 allocs/op
BenchmarkModeIterator_100K-4    	       5	 201076766 ns/op	56370464 B/op	     100 allocs/op
BenchmarkModeIterator_1M-4      	       1	2004537405 ns/op	529295120 B/op	     157 allocs/op
BenchmarkParserParseStatement-4 	  100000	     11940 ns/op	   3.77 MB/s	    7680 B/op	      72 allocs/op
BenchmarkSanitize-4             	  100000	     15126 ns/op	    1360 B/op	      11 allocs/op
BenchmarkSelect_Raw_1K-4        	  300000	      4206 ns/op	    1120 B/op	      24 allocs/op
BenchmarkSelect_Raw_100K-4      	  300000	      4158 ns/op	    1120 B/op	      24 allocs/op
BenchmarkSelect_Dedupe_1K-4     	      20	  55137127 ns/op	 4166330 B/op	  300119 allocs/op
ok  	github.com/influxdata/influxdb/influxql	31.825s
```

### CODE REVIEW

**Comment by jwilder** [comment]
@jsternberg @benbjohnson

File: influxql/call_iterator_test.go [hunk 1]
```diff
@@ -731,6 +871,60 @@ func benchmarkDistinctIterator(b *testing.B, pointN int) {
 	}
 }
 
+func BenchmarkMedianIterator_1K(b *testing.B)   { benchmarkMedianIterator(b, 1000) }
```
Comment (jsternberg) [inline influxql/call_iterator_test.go]: I really like the benchmarks for the median iterator, but is it possible for you to split this work into a separate commit and a separate PR? I would like to keep this PR as "only the implementation of `mode()`".

File: influxql/call_iterator.go [hunk 2]
```diff
@@ -495,6 +500,163 @@ func IntegerMedianReduceSlice(a []IntegerPoint) []FloatPoint {
 	return []FloatPoint{{Time: ZeroTime, Value: float64(a[len(a)/2].Value)}}
 }
 
+// newModeIterator returns an iterator for operating on a mode() call.
+func NewModeIterator(input Iterator, opt IteratorOptions) (Iterator, error) {
+
+	switch input := input.(type) {
+	case FloatIterator:
+		createFn := func() (FloatPointAggregator, FloatPointEmitter) {
+			fn := NewFloatSliceFuncReducer(FloatModeReduceSlice)
```
Comment (jsternberg) [inline influxql/call_iterator.go]: The slice func reducers are really slow and I don't think it's advisable to use them for mode since mode doesn't actually need that information. First, be sure to try this with your benchmarks to see the difference, but I would imagine that creating a custom reducer type similar to the `MeanReducer` in `functions.go` and doing something like this would be better:

```
type FloatModeReducer struct {
    m map[float64]int
}

func (r *FloatModeReducer) AggregateFloat(p *FloatPoint) {
    m[p.Value]++
}

func (r *FloatModeReducer) Emit() []FloatPoint {
    value := 0
    max := 0
    for v, count := range r.m {
        if count < max || (count == max && v < value) {
            continue
        }
        value = v
    }
}
```

I would guess that this probably ends up being faster than the slice reducer just because of much fewer memory allocations. Can you try it out?

File: influxql/call_iterator.go [hunk 3]
```diff
@@ -495,6 +500,163 @@ func IntegerMedianReduceSlice(a []IntegerPoint) []FloatPoint {
...
+		}
+		return &booleanReduceBooleanIterator{input: newBufBooleanIterator(input), opt: opt, create: createFn}, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported median iterator type: %T", input)
+	}
+}
+
+// FloatModeReduceSlice returns the mode value within a window.
+func FloatModeReduceSlice(a []FloatPoint) []FloatPoint {
```
Comment (jsternberg) [inline influxql/call_iterator.go]: These implementations are a good candidate for putting them in `functions.gen.go.tmpl` and generating the code rather than typing it out a few different times.

File: influxql/call_iterator.go (hunk 2 above)
Comment (jsternberg) [inline influxql/call_iterator.go]: I notice you also break ties later with time rather than value. I think that's fine too (and might be better) so you would just need to keep track of the first time encountered inside of the map.

File: influxql/call_iterator_test.go (hunk 1 above)
Comment (agaurav) [inline influxql/call_iterator_test.go]: makes sense, here https://github.com/influxdata/influxdb/pull/7113

File: influxql/call_iterator.go (hunk 2 above)
Comment (agaurav) [inline influxql/call_iterator.go]: I thought of this for the histogram function https://github.com/influxdata/influxdb/issues/3674 as this will give the entire frequency spectrum. The reason for opting against it for mode was that it takes an extra map of size same as the slice itself (i prefer to avoid hash map where there is a possibility for lots of uniques keys). Also, I was hoping to improve on this with the one mentioned here [sublinear-algorithm-for-calculating-the-mode-of-an-array](https://www.quora.com/What-is-the-most-efficient-algorithm-for-calculating-the-mode-of-an-array-of-integers/answer/Anders-Kaseorg) ,which needs the entire sorted slice as well.  
I will try out the custom reducer and post the result here and you can make a call then.

**Comment by agaurav** [comment]
@jsternberg I was unable to get time to implement & test the reducer based approach right now because of my own work. Can you please merge this for v1 release  ? (`mode` is kinda of essential for my implementation)
I will improve it along with histogram in a future PR.

File: influxql/call_iterator.go [hunk 4]
```diff
@@ -500,6 +500,163 @@ func IntegerMedianReduceSlice(a []IntegerPoint) []FloatPoint {
...
+	falsFreq := 0
+	mostMode := false
+
+	for _, p := range a {
+		if p.Value {
+			trueFreq++
+			continue
+		}
+		// a nil value is treated as false
+		falsFreq++
```
Comment (jsternberg) [inline influxql/call_iterator.go]: I think a nil value should probably just be skipped.

File: influxql/call_iterator.go [hunk 5]
```diff
@@ -500,6 +500,163 @@ func IntegerMedianReduceSlice(a []IntegerPoint) []FloatPoint {
...
+	mostTime := a[0].Time
+
+	for _, p := range a {
+		if p.Value != currMode {
+			currFreq = 1
+			currMode = p.Value
+			continue
+		}
+		currFreq++
+		if mostFreq > currFreq || (mostFreq == currFreq && p.Time > mostTime) {
```
Comment (jsternberg) [inline influxql/call_iterator.go]: I think the logic here for comparing times may be incorrect. I think the time should be based on the first point rather than the last point so you should track the `currTime` above when you find a new value and compare the `currTime` to `mostTime`. That way if you have the following points with 10 seconds interrupting each one:

```
2 1 1 2
```

I think this will return 1 right now. It was probably intended for this to return 2.

**Comment by jsternberg** [comment]
We're going to try and get this into 1.0 for the second release candidate. I made a few comments. Can you look over them and either implement the feedback or send me a message saying, "I agree with you, but I do not have time to implement the feedback. Can you do it for me?"

We'll also need to close this PR and reopen it targeting the 1.0 branch. Since you have signed the CLA, I can pull from your branch and do all of this for you if you want.

File: influxql/call_iterator.go (hunk 4 above)
Comment (agaurav) [inline influxql/call_iterator.go]: But wouldn't values passed to iterator will be either `true` or `false` . I made the comment assuming that it will be get all `nil` values as false.

File: influxql/call_iterator.go (hunk 5 above)
Comment (agaurav) [inline influxql/call_iterator.go]: Yeah that's wrong, great catch btw. I will correct it for the PR to v1.0 branch

**Comment by agaurav** [comment]
@jsternberg I have implemented the time time break up as per feedback. But I am getting conflicts with v1 branch, and now the pull from v1 branch has broken this PR. Can you please fix this and open the PR with your fork?

**Comment by agaurav** [comment]
You can cherry pick to till commit `ec8fa696d9600a30e75fc2c7c9e096d1e4ab45e7` or I have pushed the passing commit tree till this to branch [v1/modeFunc](https://github.com/agaurav/influxdb/tree/v1/modeFunc)

**Comment by jsternberg** [comment]
I can make the PR for you. I'm out right now but I'll do it when I get back.

File: influxql/call_iterator.go (hunk 4 above)
Comment (jsternberg) [inline influxql/call_iterator.go]: I don't actually know if it would get any nil points, but nil points would not be true or false. Nil points only happen with bare values so this query might get nil values:

```
SELECT value / total FROM cpu
```

That might get some nil values. I don't think the aggregates get them. I'm just going to remove the confusing comment.

**Comment by jsternberg** [comment]
Closing in favor of #7199.
